src/persistence/sql.rs

```diff
@@ -1,37 +1,1094 @@
+            let node = NodeSetEntry {
+                address: address.parse().context("failed to parse node address")?,
+                staking_key: staking_key.parse().context("failed to parse staking key")?,
+                state_key: state_key.parse().context("failed to parse state key")?,
+                stake: U256::from_str(&stake_str).context("failed to parse node stake")?,
+                commission: Ratio::from(commission as f32),
+            };
+            node_set.push(node);
+        }
+
+        // Load all wallets and their delegations in parallel
+        let wallet_rows =
+            sqlx::query_as::<_, (String, String)>("SELECT address, claimed_rewards FROM wallet")
+                .fetch_all(&self.pool)
+                .await?;
+
+        let wallet_futures =
+            wallet_rows
+                .into_iter()
+                .map(|(wallet_address, claimed_rewards_str)| {
+                    let pool = self.pool.clone();
+                    async move {
+                        let address: Address = wallet_address
+                            .parse()
+                            .context("failed to parse wallet address")?;
+                        let claimed_rewards = U256::from_str(&claimed_rewards_str)
+                            .context("failed to parse claimed rewards")?;
+
+                        // Get active delegations
+                        let delegation_rows = sqlx::query_as::<_, (String, String)>(
+                            "SELECT node, amount
+                             FROM delegation
+                             WHERE delegator = $1",
+                        )
+                        .bind(&wallet_address)
+                        .fetch_all(&pool)
+                        .await?;
+
+                        let mut nodes = im::Vector::new();
+                        for (node_str, amount_str) in delegation_rows {
+                            let node: Address =
+                                node_str.parse().context("failed to parse node address")?;
+                            let amount = U256::from_str(&amount_str)
+                                .context("failed to parse delegation amount")?;
+                            if !amount.is_zero() {
+                                nodes.push_back(Delegation {
+                                    delegator: address,
+                                    node,
+                                    amount,
+                                });
+                            }
+                        }
+
+                        // Get pending withdrawals
+                        let pending_rows = sqlx::query_as::<_, (String, String, String, i64)>(
+                            "SELECT node, withdrawal_type, amount, unlocks_at
+                             FROM pending_withdrawals
+                             WHERE delegator = $1
+                             ORDER BY unlocks_at ASC",
```

Comment:
Similarly, although we are ordering by `unlocks_at` here, I'm not sure it's guaranteed that this is the same order as in the snapshot that was originally stored. Might be better to represent pending withdrawals in memory as `im::OrdMap<Address, PendingWithdrawal>`, mapping node to pending withdrawal.

Actually this `ORDER BY` is not even indexed right now, so it might be better to `ORDER BY node, withdrawal_type` (which should be able to use the `(delegator, node, withdrawal_type)` index since we already have the `WHERE` clause using `delegator`)